src/apps/euler/ethereum/euler.d-token.token-fetcher.ts

```diff
@@ -0,0 +1,114 @@
+import { Network } from '~types/network.interface';
+
+import { EULER_DEFINITION } from '../euler.definition';
+
+const appId = EULER_DEFINITION.id;
+const groupId = EULER_DEFINITION.groups.dToken.id;
+const network = Network.ETHEREUM_MAINNET;
+
+const query = gql`
+  {
+    eulerMarketStore(id: "euler-market-store") {
+      markets {
+        id
+        interestRate
+        borrowAPY
+        supplyAPY
+        totalSupply
+        twap
+        name
+        symbol
+        decimals
+        dTokenAddress
+      }
+    }
+  }
+`;
+
+interface EulerMarket {
+  id: string;
+  interestRate: string;
+  borrowAPY: string;
+  supplyAPY: string;
+  totalSupply: string;
+  twap: string;
+  name: string;
+  symbol: string;
+  decimals: string;
+  dTokenAddress: string;
+}
+
+interface EulerMarketsResponse {
+  eulerMarketStore: {
+    markets: EulerMarket[];
+  };
+}
+
+@Register.TokenPositionFetcher({ appId, groupId, network })
+export class EthereumEulerDTokenTokenFetcher implements PositionFetcher<AppTokenPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(EulerContractFactory) private readonly eulerContractFactory: EulerContractFactory,
+  ) {}
+
+  async getPositions() {
+    const endpoint = 'https://api.thegraph.com/subgraphs/name/euler-xyz/euler-mainnet';
+    const data = await this.appToolkit.helpers.theGraphHelper.request<EulerMarketsResponse>({ endpoint, query });
+
+    const tokens = await Promise.all(
+      data.eulerMarketStore.markets.map(async market => {
+        const dTokenContract = this.eulerContractFactory.eulerDtokenContract({
+          address: market.dTokenAddress,
+          network,
+        });
+
+        const totalSupply = await dTokenContract.totalSupply();
+        if (totalSupply.isZero()) return null;
+
+        return {
+          address: market.dTokenAddress,
+          symbol: `D${market.symbol}`,
+          name: `Euler D token ${market.name}`,
+          type: ContractType.APP_TOKEN as const,
+          supply: Number(market.totalSupply) / 10 ** Number(market.decimals),
+          pricePerShare: Number(market.twap) / 10 ** 18,
+          price: Number(market.twap) / 10 ** 18,
+          network,
+          decimals: 18,
+          tokens: [
+            {
+              type: ContractType.BASE_TOKEN as const,
+              address: market.id,
+              symbol: market.symbol,
+              name: market.name,
+              price: Number(market.twap) / 10 ** 18,
+              network,
+              decimals: Number(market.decimals),
+            },
+          ],
+          dataProps: {
+            name: `Euler D token ${market.name}`,
+            interestRate: Number(market.interestRate) / 10 ** 18,
+            borrowAPY: Number(market.borrowAPY) / 10 ** 18,
+            supplyAPY: Number(market.borrowAPY) / 10 ** 18,
+          },
+          displayProps: { label: market.name, images: [''] },
```

Comment:
`label` should be `Euler D Token ${market.name}`, and `images` should be `getImagesFromToken(underlyingToken!)`.

You can also add a `secondaryLabel` and set that to the price of the D token like `buildDollarDisplayItem(Number(market.twap) / 10 ** 18)`